File: UCFPT/22locals/local2022FinalRoundData/solutions/wordtree/wordtree.java
// Arup Guha
// 8/9/2022
// Solution to 2022 UCF Locals Problem: Word Tree

import java.util.*;

public class wordtree {
	
	public static int n;
	public static int k;
	public static char[][] words;
	
	public static void main(String[] args) {
		
		// Read the input.
		Scanner stdin = new Scanner(System.in);
		n = stdin.nextInt();
		k = stdin.nextInt();
		words = new char[n][];
		for (int i=0; i<n; i++)
			words[i] = stdin.next().toCharArray();
			
		// Create all possible edges.
		ArrayList<edge> eList = new ArrayList<edge>();
		for (int i=0; i<n; i++)
			for (int j=i+1; j<n; j++)
				eList.add(new edge(i, j, cost(i,j)));
				
		// Sort for Kruskal's.
		Collections.sort(eList);
		int numE = 0;
		
		// Set up our disjoint set.
		djset dj = new djset(n);
		int idx = 0, res = -1;
		
		while (numE < n-1) {
		
			// Get next edge, skip if can't use.
			edge e = eList.get(idx++);
			boolean fuse = dj.union(e.u, e.v);
			if (!fuse) continue;
			
			// New edge add it.
			numE++;
			
			// If we always assign, max edge weight will be correct at end.
			res = e.w;
		}
		
		// Ta da!
		System.out.println(res);
	}
	
	// Returns the edge cost between words i and j.
	public static int cost(int i, int j) {
		int res = 0;
		for (int z=0; z<k; z++)
			res += Math.abs(words[i][z]-words[j][z]);
		return res;
	}

}

class edge implements Comparable<edge> {
	
	public int u;
	public int v;
	public int w;
	
	public edge(int myu, int myv, int myw) {
		u = myu;
		v = myv;
		w = myw;
	}
	
	public int compareTo(edge other) {
		return this.w - other.w;
	}
}

// Disjoint set class with path compression.
class djset {

	public int n;
	public int[] par;
	
	public djset(int myn) {
		n = myn;
		par = new int[n];
		for (int i=0; i<n; i++)
			par[i] = i;
	}
	
	public int find(int v) {
		if (par[v] == v) return v;
		return par[v] = find(par[v]);
	}
	
	public boolean union(int u, int v) {
		u = find(u);
		v = find(v);
		if (u == v) return false;
		par[v] = u;
		return true;
	}
}

File: UCFPT/22locals/local2022FinalRoundData/solutions/wordtree/wordtree.cpp
// Arup Guha
// 9/6/2022
// C++ Solution to 2022 UCF Locals Qualifying Problem: Word Tree

using namespace std;

#include <iostream>
#include <queue>
#include <string>
#include <cmath>

#define MAX(a,b) ((a) > (b) ? a : b)

struct edge {
    int u,v,cost;
	
    bool operator<(const edge& other) const {
        return other.cost < cost;
    }
};

// Other globals I use.
int n, k;
string* words;

// Function prototypes.
int cost(int i, int j);

int main(void) {
	
	cin >> n >> k;
	words = new string[n];
	for (int i=0; i<n; i++)
		cin >> words[i];
		
	// Set up visited array for Prim's
	bool* used = new bool[n];
	for (int i=0; i<n; i++) used[i] = false;
	used[0] = true;
	
	// Set up PQ for Prim's.
	priority_queue<edge> pq;
	for (int i=1; i<n; i++) pq.push(edge{0,i,cost(0, i)});	
	
	// Will store our result.
	int res = 0, numV = 1;
	
	// While we're not connected.
	while (numV < n) {
	
		edge cur = pq.top();
		pq.pop();
		
		// Not to a new place.
		if (used[cur.u] && used[cur.v]) continue;
		
		// The new place.
		int newV = used[cur.u] ? cur.v : cur.u;
	
		// Bookkeeping - update result, used vertices & num vertices.
		res = MAX(res, cur.cost);
		used[newV] = true;
		numV++;
		
		// Add all edges incident to newV to pq.
		for (int i=0; i<n; i++) {
			if (i == newV) continue;
			pq.push(edge{newV,i,cost(newV,i)});
		}
	}
	
	// Ta da!
	cout << res << endl;
	
	// Clean up!
	delete [] words;
	delete [] used;
	return 0;
}

// Returns the cost between strings i and j.
int cost(int i, int j) {
	int res = 0;
	for (int z=0; z<k; z++)
		res += abs(words[i][z]-words[j][z]);
	return res;
}
